File: backend/inspections/services/scoring.py
from inspections.models import ChecklistItem, ChecklistSection, Inspection, InspectionAnswer


def compute_section_scores(inspection: Inspection) -> dict:
    sections = ChecklistSection.objects.prefetch_related("items").order_by("order")
    if inspection.checklist_version_id:
        sections = sections.filter(version_id=inspection.checklist_version_id)
    answers_by_item = {a.checklist_item_id: a for a in inspection.answers.select_related("checklist_item")}

    section_scores = []
    total_c = total_nc = total_na = 0

    for section in sections:
        c = nc = na = 0
        for item in section.items.all():
            answer = answers_by_item.get(item.id)
            if not answer or not answer.status:
                continue
            if answer.status == InspectionAnswer.AnswerStatus.C:
                c += 1
            elif answer.status == InspectionAnswer.AnswerStatus.NC:
                nc += 1
            elif answer.status == InspectionAnswer.AnswerStatus.NA:
                na += 1

        applicable = c + nc + na
        score = c / applicable if applicable > 0 else None
        total_c += c
        total_nc += nc
        total_na += na

        section_scores.append(
            {
                "section_id": section.id,
                "section_order": section.order,
                "section_title": section.title,
                "conforme": c,
                "nao_conforme": nc,
                "nao_aplicavel": na,
                "total_applicable": applicable,
                "score": score,
            }
        )

    overall_applicable = total_c + total_nc + total_na
    overall_score = total_c / overall_applicable if overall_applicable > 0 else None

    return {
        "sections": section_scores,
        "overall_conforme": total_c,
        "overall_nao_conforme": total_nc,
        "overall_nao_aplicavel": total_na,
        "overall_score": overall_score,
    }


def get_non_conformities(inspection: Inspection) -> list[dict]:
    nc_items = []
    answers = inspection.answers.select_related("checklist_item__section").prefetch_related("photos")
    for answer in answers:
        if answer.status != InspectionAnswer.AnswerStatus.NC:
            continue
        item = answer.checklist_item
        nc_items.append(
            {
                "item_code": item.item_code,
                "question": item.question,
                "description": answer.description or item.question,
                "normative": answer.normative,
                "recommendation": answer.recommendation,
                "section_title": item.section.title,
                "photos": [p.image.path for p in answer.photos.all() if p.image],
            }
        )
    nc_items.sort(key=lambda x: [int(p) if p.isdigit() else p for p in x["item_code"].replace(".", " ").split()])
    return nc_items


def get_checklist_summary(inspection: Inspection) -> list[dict]:
    """Seções do checklist com todos os itens que receberam resposta (C, NC ou NA)."""
    sections = ChecklistSection.objects.prefetch_related("items").order_by("order")
    if inspection.checklist_version_id:
        sections = sections.filter(version_id=inspection.checklist_version_id)

    answers_by_item = {a.checklist_item_id: a for a in inspection.answers.select_related("checklist_item")}

    status_labels = {
        InspectionAnswer.AnswerStatus.C: "Conforme (C)",
        InspectionAnswer.AnswerStatus.NC: "Não conforme (NC)",
        InspectionAnswer.AnswerStatus.NA: "Não aplicável (NA)",
    }

    summary = []
    for section in sections:
        items = []
        for item in section.items.all():
            answer = answers_by_item.get(item.id)
            if not answer or not answer.status:
                continue
            items.append(
                {
                    "item_code": item.item_code,
                    "question": item.question,
                    "status": answer.status,
                    "status_label": status_labels.get(answer.status, answer.status),
                    "description": answer.description or "",
                }
            )
        if items:
            summary.append(
                {
                    "section_order": section.order,
                    "section_title": section.title,
                    "checklist_items": items,
                }
            )
    return summary

File: frontend/src/lib/report/scoring.ts
import type { ChecklistSection, SsmaConfig } from "@/lib/api";
import type { LocalAnswer, LocalInspection, LocalPhoto } from "@/lib/db";
import type { UnitInput } from "@/lib/api";

export type SectionScore = {
  section_id: number;
  section_order: number;
  section_title: string;
  conforme: number;
  nao_conforme: number;
  nao_aplicavel: number;
  total_applicable: number;
  score: number | null;
};

export type ScoresResult = {
  sections: SectionScore[];
  overall_conforme: number;
  overall_nao_conforme: number;
  overall_nao_aplicavel: number;
  overall_score: number | null;
};

export function computeSectionScores(
  sections: ChecklistSection[],
  answers: LocalAnswer[]
): ScoresResult {
  const byItem = new Map(answers.map((a) => [a.checklist_item_id, a]));
  const sectionScores: SectionScore[] = [];
  let totalC = 0;
  let totalNc = 0;
  let totalNa = 0;

  for (const section of sections) {
    let c = 0;
    let nc = 0;
    let na = 0;
    for (const item of section.items) {
      const answer = byItem.get(item.id);
      if (!answer?.status) continue;
      if (answer.status === "C") c += 1;
      else if (answer.status === "NC") nc += 1;
      else if (answer.status === "NA") na += 1;
    }
    const applicable = c + nc + na;
    sectionScores.push({
      section_id: section.id,
      section_order: section.order,
      section_title: section.title,
      conforme: c,
      nao_conforme: nc,
      nao_aplicavel: na,
      total_applicable: applicable,
      score: applicable > 0 ? c / applicable : null,
    });
    totalC += c;
    totalNc += nc;
    totalNa += na;
  }

  const overallApplicable = totalC + totalNc + totalNa;
  return {
    sections: sectionScores,
    overall_conforme: totalC,
    overall_nao_conforme: totalNc,
    overall_nao_aplicavel: totalNa,
    overall_score: overallApplicable > 0 ? totalC / overallApplicable : null,
  };
}

export type NonConformity = {
  item_code: string;
  question: string;
  description: string;
  normative: string;
  recommendation: string;
  section_title: string;
  photoDataUris: string[];
};

export function getNonConformities(
  sections: ChecklistSection[],
  answers: LocalAnswer[],
  photos: LocalPhoto[],
  blobToDataUri: (blob: Blob) => Promise<string>
): Promise<NonConformity[]> {
  const byItem = new Map(answers.map((a) => [a.checklist_item_id, a]));
  const photosByItem = new Map<number, LocalPhoto[]>();
  for (const p of photos.filter((x) => x.photo_type === "nc")) {
    const list = photosByItem.get(p.checklist_item_id) ?? [];
    list.push(p);
    photosByItem.set(p.checklist_item_id, list);
  }

  const tasks: Promise<NonConformity>[] = [];
  for (const section of sections) {
    for (const item of section.items) {
      const answer = byItem.get(item.id);
      if (answer?.status !== "NC") continue;
      tasks.push(
        (async () => {
          const itemPhotos = photosByItem.get(item.id) ?? [];
          const photoDataUris: string[] = [];
          for (const ph of itemPhotos) {
            photoDataUris.push(await blobToDataUri(ph.blob));
          }
          return {
            item_code: item.item_code,
            question: item.question,
            description: answer.description || item.question,
            normative: answer.normative,
            recommendation: answer.recommendation,
            section_title: section.title,
            photoDataUris,
          };
        })()
      );
    }
  }

  return Promise.all(tasks).then((items) =>
    items.sort((a, b) => a.item_code.localeCompare(b.item_code, undefined, { numeric: true }))
  );
}

export function inspectionCoverConfig(
  inspection: LocalInspection,
  unit: UnitInput,
  ssma: SsmaConfig
) {
  return {
    diretoria_executiva: ssma.diretoria_executiva,
    diretor_executivo: inspection.cover_diretor_executivo || ssma.diretor_executivo,
    gerencia_geral: ssma.gerencia_geral,
    gerente_geral: inspection.cover_gerente_geral || ssma.gerente_geral,
    gerencia_sst: ssma.gerencia_sst,
    gerente_sst: inspection.cover_gerente_sst || ssma.gerente_sst,
    gerencia_meio_ambiente: ssma.gerencia_meio_ambiente,
    gerente_meio_ambiente: inspection.cover_gerente_meio_ambiente || ssma.gerente_meio_ambiente,
    regional: unit.regional || ssma.regional,
    cidade: unit.city || ssma.cidade,
  };
}

function fmtScore(score: number | null): string {
  if (score == null) return "—";
  return `${(score * 100).toFixed(1).replace(".", ",")}%`;
}

export function fmtPct(score: number | null): string {
  return fmtScore(score);
}
